File: front/code-book/src/types/code.ts
import { Comment } from './comment'
import { Tag } from './tag'
import { User } from './user'

export type Code = FileCode | UrlCode | SnippetCode

type CodeBase = {
  id: string
  title: string
  author: User
  created: Date
  tags: Tag[]
  comments: Comment[]
}

type File = {
  name: string
  text: string
}

export type FileCode = CodeBase & {
  kind: 'file'
  files: File[]
}

export type UrlCode = CodeBase & {
  kind: 'url'
  url: string
}

export type Variable = {
  name: string
  value: string
}

export type SnippetCode = CodeBase & {
  kind: 'snippet'
  file: File
  variables: Variable[]
}

File: back/code-book/src/main/java/com/codebook/codebook/code/Code.java
package com.codebook.codebook.code;

import com.codebook.codebook.comment.Comment;
import com.codebook.codebook.user.User;
import lombok.Value;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Value
public class Code {

    String id;

    String title;

    User author;

    String created;

    List<Comment> comments;

    CodePart part;

    public Map<String, Object> toMap() {
        return Map.of(
            "id", id,
            "title", title,
            "author", author.toMap(),
            "created", created,
            "part", part.toMap(),
            "tags", part.getTags(),
            "comments", comments.stream().map(Comment::toMap).collect(Collectors.toList())
        );
    }

}
